Nudging breaks sync messages with an execution

This issue happens when #371 is merged. Still, I wanted to open a new issue on that as I think there is more involved here.

Given the following scenario:
![image](https://user-images.githubusercontent.com/588090/46410966-cac84600-c71a-11e8-80cf-bf7657d467f9.png)

When we now add a new sync message after the existing reply message from `Lifeline2` to `Lifeline1` the following happens:
![image](https://user-images.githubusercontent.com/588090/46411058-167aef80-c71b-11e8-9262-4319025c9c5d.png)

This wrong nudging behavior seems to be caused by `InsertMessageCommand.createNudgeCommandForFollowingElements(...)`, in particular this code:
```java
MElement<?> distanceFrom = (MElement<?>)Optional.of(beforeSend).filter(MOccurrence.class::isInstance)
	.map(MOccurrence.class::cast).flatMap(MOccurrence::getStartedExecution).orElse(beforeSend);
Optional<Command> makeSpace = getFollowingElement(timeline, distanceFrom, sendYPosition).map(el -> {
	OptionalInt distance = el.verticalDistance(distanceFrom);
	if (distance.isPresent()) {
		return el.nudge(Math.max(minNudge, spaceRequired - distance.getAsInt()));
	} else {
		return null;
	}
});
```

`distanceFrom` is the send event of the existing reply message. This send event is correctly sorted in _before_ the receive event of the reply message in the `timeline`. Consequently, `getFollowingElement(...)` will return and invoke nudging on the receive event of the reply message, which is wrong, as this reply event must not be nudged.

My approach for solving this issue is to find the latest (in the sense of the timeline) element that is before the y-position of the inserted message. This way, the receive event wouldn't be nudged anymore, only elements that occur _after_ graphically and time-wise would be nudged. However, this breaks the `SelfMessageReconnectionUITest`, and I'm not sure I really understand the rationale of these tests to know if this is indeed a negative side-effect or what I broke:
```[synchCall, RECEIVE_ON_OTHER_LIFELINE]
reconnectSelfMessage[synchCall, RECEIVE_ON_OTHER_LIFELINE]
java.lang.AssertionError: Lifeline and slope incorrect
Expected: points point list runs from point that x that 121 ± 2 and y that 195 ± 2 to point that x that 276 ± 2 and y that 195 ± 2
     but: points x was <120>
```

```[asynchCall, RECEIVE_ON_OTHER_LIFELINE]
reconnectSelfMessage[asynchCall, RECEIVE_ON_OTHER_LIFELINE]
java.lang.AssertionError: Lifeline and slope incorrect
Expected: points point list runs from point that x that 121 ± 2 and y that 195 ± 2 to point that x that 276 ± 2 and y that 215 ± 2
     but: points x was <120>
```

I'll push a separate PR to provide my initial attempt to solve this problem.

--- a/plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/commands/InsertMessageCommand.java
+++ b/plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/commands/InsertMessageCommand.java
@@ -737,9 +737,19 @@ public class InsertMessageCommand extends ModelCommand<MLifelineImpl> implements
 		MElement<?> latestElementBeforeY = getLatestElementBeforeY(beforeSend, sendYPosition, timeline);
 		Optional<Command> makeSpace = getFollowingElement(timeline, latestElementBeforeY, sendYPosition)
 				.map(el -> {
-					OptionalInt distance = el.verticalDistance(latestElementBeforeY);
+					MElement<? extends Element> toNudge = el;
+					if (el instanceof MExecution) {
+						// We are inserting within the vertical span of an execution
+						// specification. Don't nudge the top of the execution, but
+						// rather the bottom (the finish occurrence), to make space
+						MExecution exec = (MExecution)el;
+						if (exec.getFinish().isPresent()) {
+							toNudge = exec.getFinish().get();
+						}
+					}
+					OptionalInt distance = toNudge.verticalDistance(latestElementBeforeY);
 					if (distance.isPresent()) {
-						return el.nudge(Math.max(minNudge, spaceRequired - distance.getAsInt()));
+						return toNudge.nudge(Math.max(minNudge, spaceRequired - distance.getAsInt()));
 					} else {
 						return null;
 					}

